sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2178,25 +2179,87 @@ std::pair<RT::PiKernel, std::mutex *> ProgramManager::getOrCreateKernel(
-      Img.getDeviceRequirements();
-  RTDeviceBinaryImage::PropertyRange::ConstIterator PropIt = std::find_if(
-      PropRange.begin(), PropRange.end(),
-      [](RTDeviceBinaryImage::PropertyRange::ConstIterator &&Prop) {
-        using namespace std::literals;
-        return (*Prop)->Name == "aspects"sv;
-      });
-  if (PropIt == PropRange.end())
+bool doesDevSupportDeviceRequirements(const device &Dev,
+                                      const RTDeviceBinaryImage &Img) {
+  auto getPropIt = [&Img](const std::string &PropName) {
+    const RTDeviceBinaryImage::PropertyRange &PropRange =
+        Img.getDeviceRequirements();
+    RTDeviceBinaryImage::PropertyRange::ConstIterator PropIt = std::find_if(
+        PropRange.begin(), PropRange.end(),
+        [&PropName](RTDeviceBinaryImage::PropertyRange::ConstIterator &&Prop) {
+          return (*Prop)->Name == PropName;
+        });
+    return (PropIt == PropRange.end())
+               ? std::nullopt
+               : std::optional<
+                     RTDeviceBinaryImage::PropertyRange::ConstIterator>{PropIt};
+  };
+
+  auto AspectsPropIt = getPropIt("aspects");
+  auto ReqdWGSizePropIt = getPropIt("reqd_work_group_size");
+
+  if (!AspectsPropIt && !ReqdWGSizePropIt)
     return true;
-  ByteArray Aspects = DeviceBinaryProperty(*PropIt).asByteArray();
-  // Drop 8 bytes describing the size of the byte array.
-  Aspects.dropBytes(8);
-  while (!Aspects.empty()) {
-    aspect Aspect = Aspects.consume<aspect>();
-    if (!Dev.has(Aspect))
+
+  // Checking if device supports defined aspects
+  if (AspectsPropIt) {
+    ByteArray Aspects =
+        DeviceBinaryProperty(*(AspectsPropIt.value())).asByteArray();
+    // Drop 8 bytes describing the size of the byte array.
+    Aspects.dropBytes(8);
+    while (!Aspects.empty()) {
+      aspect Aspect = Aspects.consume<aspect>();
+      if (!Dev.has(Aspect))
+        return false;
+    }
+  }
+
+  // Checking if device supports defined required work group size
+  if (ReqdWGSizePropIt) {
+    ByteArray ReqdWGSize =
+        DeviceBinaryProperty(*(ReqdWGSizePropIt.value())).asByteArray();
+    // Drop 8 bytes describing the size of the byte array.
+    ReqdWGSize.dropBytes(8);
+    int ReqdWGSizeAllDimsTotal = 1;
+    std::vector<int> ReqdWGSizeVec;
+    int Dims = 0;
+    while (!ReqdWGSize.empty()) {
+      int SingleDimSize = ReqdWGSize.consume<int>();
+      ReqdWGSizeAllDimsTotal *= SingleDimSize;
+      ReqdWGSizeVec.push_back(SingleDimSize);
+      Dims++;
+    }
+    if (static_cast<long unsigned int>(ReqdWGSizeAllDimsTotal) >
+        Dev.get_info<info::device::max_work_group_size>())
       return false;
+    // Creating std::variant to call max_work_item_sizes one time to avoid
+    // performance drop
+    std::variant<id<1>, id<2>, id<3>> MaxWorkItemSizesVariant;
+    if (Dims == 1)
+      MaxWorkItemSizesVariant =
+          Dev.get_info<info::device::max_work_item_sizes<1>>();
```

Comment:
@steffenlarsen by querying `max_work_item_sizes<1>`, `max_work_item_sizes<2>` and `max_work_item_sizes<3>` I assumed that we can query different numbers depending on dimension, e.g., for 1-dim, there will be {A} and for 2-dim there will be {B, C}. This is because I didn't find anything related to this in OpenCL/L0/SYCL specs. What do you think? If different dimensions return same numbers (1D: {A}, 2D: {A, B}, 3D: {A, B, C}) we of course can call only `max_work_item_sizes<3>` and just truncate for the required dimension.

Reply:
To my knowledge we do not require that the value in `max_work_item_sizes<1>` is in any of the dimensions of `max_work_item_sizes<2>`, so I believe what you have here is the safest solution. Looks good! 😄

Reply:
OK, thanks!